cmd/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,327 @@
+	if beta {
+		siteURL = "https://beta.rclone.org"
+	}
+	newVersion = version
+	if newVersion == "" {
+		_, newVersion, _, err = versionCmd.GetVersion(siteURL + "/version.txt")
+	}
+	return
+}
+
+// InstallUpdate performs rclone self-update
+func InstallUpdate(opt *Options) error {
+	// Find the latest release number
+	if opt.Stable && opt.Beta {
+		return errors.New("--stable and --beta are mutually exclusive")
+	}
+
+	newVersion, siteURL, err := GetVersion(opt.Beta, opt.Version)
+	if err != nil {
+		return errors.Wrap(err, "unable to detect new version")
+	}
+
+	if newVersion == "" {
+		var err error
+		_, newVersion, _, err = versionCmd.GetVersion(siteURL + "/version.txt")
+		if err != nil {
+			return errors.Wrap(err, "unable to detect new version")
+		}
+	}
+
+	if newVersion == fs.Version {
+		fmt.Println("rclone is up to date")
+		return nil
+	}
+
+	// Find the executable path
+	executable, err := os.Executable()
+	if err != nil {
+		return errors.Wrap(err, "unable to find executable")
+	}
+
+	targetFile := opt.Output
+	if targetFile == "" {
+		targetFile = executable
+	}
+
+	// Check for possible access errors in advance
+	newFile := targetFile + ".new"
+	savedFile := ""
+	if runtime.GOOS == "windows" {
+		savedFile = targetFile + ".old"
+	}
+
+	if savedFile == executable || newFile == executable {
+		return fmt.Errorf("%s: a temporary file would overwrite executable, specify a different --output", targetFile)
+	}
+
+	if err := verifyAccess(targetFile); err != nil {
+		return err
+	}
+
+	// Download the update as a temporary file
+	if err := downloadUpdate(opt.Beta, newVersion, siteURL, newFile); err != nil {
+		return errors.Wrap(err, "failed to update rclone")
+	}
+
+	// Copy permission bits from the old executable
+	fileMode := os.FileMode(0755)
+	if fileInfo, err := os.Lstat(targetFile); err == nil {
+		fileMode = fileInfo.Mode()
+	}
+	if err := os.Chmod(newFile, fileMode); err != nil {
+		return errors.Wrap(err, "failed to set permissions")
+	}
+
+	// Replace current executable by the new file
+	if err = os.Remove(targetFile); os.IsNotExist(err) {
+		err = nil
+	}
+	if err != nil && savedFile != "" {
+		// Windows forbids removal of a running executable so we rename it
+		var saveErr error
+		if saveErr = os.Remove(savedFile); os.IsNotExist(saveErr) {
+			saveErr = nil
+		}
+		if saveErr == nil {
+			saveErr = os.Rename(targetFile, savedFile)
+		}
+		if saveErr == nil {
+			fmt.Printf("The old executable was saved as %s\n", savedFile)
+			err = nil
+		} else {
+			// The rename trick didn't work out, proceed like on Unix
+			_ = os.Remove(savedFile)
+		}
+	}
+	if err == nil {
+		err = os.Rename(newFile, targetFile)
+	}
+	if err != nil {
+		_ = os.Remove(newFile)
+		return err
+	}
+	fmt.Printf("Successfully updated rclone to version %s\n", newVersion)
+	return nil
+}
+
+func downloadUpdate(beta bool, version, siteURL, newFile string) error {
+	archiveFilename := fmt.Sprintf("rclone-%s-%s-%s.zip", version, runtime.GOOS, runtime.GOARCH)
```

Comment:
Note that for GOOS="darwin" we rename that to "osx" since no mac user knows their machines run darwin.

(This should probably be `macOS` now-a-days in the age of macOS 11)